[PyTorch] Lazily construct guts of RecordFunction

swolchok opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #48286 [PyTorch] Inline RecordFunctionCallback::shouldRun
* **#47550 [PyTorch] Lazily construct guts of RecordFunction**
* #48274 [PyTorch] Make RecordFunctionCallback::should_run_ a function pointer

I saw over 5% time spent in RecordFunction's ctor during one
of our framework overhead benchmarks in `perf`. Inspecting assembly,
it looks like we just create a lot of RecordFunctions and the
constructor has to initialize a relatively large number of member
variables.

This diff takes advantage of the observation that RecordFunction does
nothing most of the time by moving its state onto the heap and only
allocating it if needed. It does add the requirement that profiling is
actually active to use RecordFunction accessors, which I hope won't be
a problem.

Differential Revision: [D24812213](https://our.internmc.facebook.com/intern/diff/D24812213/)

swolchok committed:
[PyTorch] Lazily construct guts of RecordFunction

I saw over 5% time spent in RecordFunction's ctor during one
of our framework overhead benchmarks in `perf`. Inspecting assembly,
it looks like we just create a lot of RecordFunctions and the
constructor has to initialize a relatively large number of member
variables.

This diff takes advantage of the observation that RecordFunction does
nothing most of the time by moving its state onto the heap and only
allocating it if needed. It does add the requirement that profiling is
actually active to use RecordFunction accessors, which I hope won't be
a problem.

Differential Revision: [D24812213](https://our.internmc.facebook.com/intern/diff/D24812213/)

[ghstack-poisoned]

dzhulgakov commented on aten/src/ATen/record_function.h:
```diff
@@ -174,75 +187,91 @@ struct TORCH_API RecordFunction {
 
   inline RecordFunctionHandle handle() const {
-    return handle_;
+    TORCH_CHECK(state_, "Called handle() on inactive RecordFunction");
```
we saw in the past that TORCH_CHECK in inline functions can be costly (as it generates a full blown if statement and can throw the compiler's inline logic off the tracks). I'd try to replace it with TORCH_INTERNAL_ASSERT_DEBUG_ONLY or replace it all together to see if you get further savings

btw, TORCH_CHECK -> TORCH_INTERNAL_ASSERT anyway. They are kind of similar but generate different error messages: TORCH_CHECK is more of a user error, INTERNAL_ASSERT is for PyTorch invariant violations (which can happen only because of code bugs)

dzhulgakov reviewed (commented):
Awesome!

swolchok committed:
s/TORCH_CHECK/TORCH_INTERNAL_ASSERT_DEBUG_ONLY on "[PyTorch] Lazily construct guts of RecordFunction"


I saw over 5% time spent in RecordFunction's ctor during one
of our framework overhead benchmarks in `perf`. Inspecting assembly,
it looks like we just create a lot of RecordFunctions and the
constructor has to initialize a relatively large number of member
variables.

This diff takes advantage of the observation that RecordFunction does
nothing most of the time by moving its state onto the heap and only
allocating it if needed. It does add the requirement that profiling is
actually active to use RecordFunction accessors, which I hope won't be
a problem.

Differential Revision: [D24812213](https://our.internmc.facebook.com/intern/diff/D24812213/)

[ghstack-poisoned]

swolchok committed:
revert stray TORCH_CHECK removal on "[PyTorch] Lazily construct guts of RecordFunction"


I saw over 5% time spent in RecordFunction's ctor during one
of our framework overhead benchmarks in `perf`. Inspecting assembly,
it looks like we just create a lot of RecordFunctions and the
constructor has to initialize a relatively large number of member
variables.

This diff takes advantage of the observation that RecordFunction does
nothing most of the time by moving its state onto the heap and only
allocating it if needed. It does add the requirement that profiling is
actually active to use RecordFunction accessors, which I hope won't be
a problem.

Differential Revision: [D24812213](https://our.internmc.facebook.com/intern/diff/D24812213/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Lazily construct guts of RecordFunction"


I saw over 5% time spent in RecordFunction's ctor during one
of our framework overhead benchmarks in `perf`. Inspecting assembly,
it looks like we just create a lot of RecordFunctions and the
constructor has to initialize a relatively large number of member
variables.

This diff takes advantage of the observation that RecordFunction does
nothing most of the time by moving its state onto the heap and only
allocating it if needed. It does add the requirement that profiling is
actually active to use RecordFunction accessors, which I hope won't be
a problem.

Differential Revision: [D24812213](https://our.internmc.facebook.com/intern/diff/D24812213/)

[ghstack-poisoned]

swolchok committed:
only activate if found_active_cb on "[PyTorch] Lazily construct guts of RecordFunction"


I saw over 5% time spent in RecordFunction's ctor during one
of our framework overhead benchmarks in `perf`. Inspecting assembly,
it looks like we just create a lot of RecordFunctions and the
constructor has to initialize a relatively large number of member
variables.

This diff takes advantage of the observation that RecordFunction does
nothing most of the time by moving its state onto the heap and only
allocating it if needed. It does add the requirement that profiling is
actually active to use RecordFunction accessors, which I hope won't be
a problem.

Differential Revision: [D24812213](https://our.internmc.facebook.com/intern/diff/D24812213/)

[ghstack-poisoned]

swolchok committed:
actually run init_handles, whoops on "[PyTorch] Lazily construct guts of RecordFunction"


I saw over 5% time spent in RecordFunction's ctor during one
of our framework overhead benchmarks in `perf`. Inspecting assembly,
it looks like we just create a lot of RecordFunctions and the
constructor has to initialize a relatively large number of member
variables.

This diff takes advantage of the observation that RecordFunction does
nothing most of the time by moving its state onto the heap and only
allocating it if needed. It does add the requirement that profiling is
actually active to use RecordFunction accessors, which I hope won't be
a problem.

Differential Revision: [D24812213](https://our.internmc.facebook.com/intern/diff/D24812213/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Lazily construct guts of RecordFunction"


I saw over 5% time spent in RecordFunction's ctor during one
of our framework overhead benchmarks in `perf`. Inspecting assembly,
it looks like we just create a lot of RecordFunctions and the
constructor has to initialize a relatively large number of member
variables.

This diff takes advantage of the observation that RecordFunction does
nothing most of the time by moving its state onto the heap and only
allocating it if needed. It does add the requirement that profiling is
actually active to use RecordFunction accessors, which I hope won't be
a problem.

Differential Revision: [D24812213](https://our.internmc.facebook.com/intern/diff/D24812213/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Lazily construct guts of RecordFunction"


I saw over 5% time spent in RecordFunction's ctor during one
of our framework overhead benchmarks in `perf`. Inspecting assembly,
it looks like we just create a lot of RecordFunctions and the
constructor has to initialize a relatively large number of member
variables.

This diff takes advantage of the observation that RecordFunction does
nothing most of the time by moving its state onto the heap and only
allocating it if needed. It does add the requirement that profiling is
actually active to use RecordFunction accessors, which I hope won't be
a problem.

Differential Revision: [D24812213](https://our.internmc.facebook.com/intern/diff/D24812213/)

[ghstack-poisoned]

swolchok commented:
Can I get a review on this one, please?

ezyang commented:
@ilia-cher I can review this one in a pinch, but you should really take a look here.

swolchok committed:
remove dead vectors; call manager() only if needed on "[PyTorch] Lazily construct guts of RecordFunction"


I saw over 5% time spent in RecordFunction's ctor during one
of our framework overhead benchmarks in `perf`. Inspecting assembly,
it looks like we just create a lot of RecordFunctions and the
constructor has to initialize a relatively large number of member
variables.

This diff takes advantage of the observation that RecordFunction does
nothing most of the time by moving its state onto the heap and only
allocating it if needed. It does add the requirement that profiling is
actually active to use RecordFunction accessors, which I hope won't be
a problem.

Differential Revision: [D24812213](https://our.internmc.facebook.com/intern/diff/D24812213/)

[ghstack-poisoned]

swolchok committed:
go ahead and allocate state if there are callbacks on "[PyTorch] Lazily construct guts of RecordFunction"


I saw over 5% time spent in RecordFunction's ctor during one
of our framework overhead benchmarks in `perf`. Inspecting assembly,
it looks like we just create a lot of RecordFunctions and the
constructor has to initialize a relatively large number of member
variables.

This diff takes advantage of the observation that RecordFunction does
nothing most of the time by moving its state onto the heap and only
allocating it if needed. It does add the requirement that profiling is
actually active to use RecordFunction accessors, which I hope won't be
a problem.

Differential Revision: [D24812213](https://our.internmc.facebook.com/intern/diff/D24812213/)

[ghstack-poisoned]

swolchok committed:
expect inactive callbacks. fold state allocation into init_handles to avoid temporary storage. on "[PyTorch] Lazily construct guts of RecordFunction"


I saw over 5% time spent in RecordFunction's ctor during one
of our framework overhead benchmarks in `perf`. Inspecting assembly,
it looks like we just create a lot of RecordFunctions and the
constructor has to initialize a relatively large number of member
variables.

This diff takes advantage of the observation that RecordFunction does
nothing most of the time by moving its state onto the heap and only
allocating it if needed. It does add the requirement that profiling is
actually active to use RecordFunction accessors, which I hope won't be
a problem.

Differential Revision: [D24812213](https://our.internmc.facebook.com/intern/diff/D24812213/)

[ghstack-poisoned]
